h/zis/client.h

```diff
@@ -422,6 +422,7 @@ int zisCallNWMService(const CrossMemoryServerName *serverName,
 
 int zisGenerateOrValidateSafIdt(const CrossMemoryServerName *serverName,
```

Comment:
The h/zis/client.h and c/zis/client.c sources are parts of zss build script. In runtime, `AuthServiceParmList` is the interface between zss and zis server. The parmlist is created and set in the zss and now contains the `options` flag byte to maintain compatibility between different versions of zss and zis server. 

- If a new zss (supporting SAF IDT/ APPL keyword) talks to an old zis server then new parameters are not visible to the old zis server.

- if an old zss server talks to the new zis server then the flag byte is set to zeros and zis server takes the old path (The flag byte was created from the reserved byte in the parmlist and the parmlist storage is zeroed after it is allocated).

I tried to design it in a way that if two different versions of zss and zis server meet then the zis server will not crash or return unpredictable results.

To answer the concerns above:

- _Also, how would the previous version of this ZIS service behave if it was called with ZIS_AUTH_SERVICE_PARMLIST_OPTION_IDT_APPL?_
The zis server would ignore the parm because it can't see it.

- _Will it always fail or will it return some unexpected results?_
It should not fail and should return the results like no APPL keyword is specified. 
From the user perspective, it is not ideal because a user would get back a JWT token that contains *ANYAPPL* claim indicating no APPL keyword was specified.

Reply:
Ok, I haven't thought of the ZSS plugin case. I'll create a new function for the appl keyword then. 

If it's a bug? I'm not sure. I think it would be a bug if it's generally OK to run zss with zis server of a different version. Is there any statement that would say what to expect if zss and zis are of different versions? Can one expect a proper behavior in such a case?

Reply:
Yes, it's absolutely possible to run into situations where 2 are of different versions. For one, you can have multiple ZIS instances and multiple ZSS instances, you can also re-configure your ZSS to talk to a different ZIS.

I think the behavior should be explicit, especially one related to security.

Reply:
Ok, I see. Can you please advise how to technically resolve it? I think the security request should be discarded in zss before it reaches an old zis server because once in the zis server, the server is not aware of the APPL keyword.

Reply:
Sorry to keep you waiting. Not sure if I have an answer. I can see issues in the original change: the parmlist got expanded in a non-backward compatible way. I need to think...

Reply:
I reverted the change to the public function.

Regarding the current APPL backward compatibility, I think a new function code could be defined that will result in an error for the older ZIS server:
https://github.com/zowe/zss/blob/staging/h/zis/services/auth.h#L50-L53

Reply:
This PR also seems to have a bug: https://github.com/zowe/zss/pull/370. What happens to that new functionality if ZSS uses the previous version of ZIS?

As for these changes, why don't we make a plugin? We should probably stop adding functionality in ZIS itself and just extended using plugins. It's apparent that the auth service doesn't really have means to check versions of its parmlist which can lead to security bugs between different versions.

CC: @1000TurquoisePogs

Reply:
> This PR also seems to have a bug: #370. What happens to that new functionality if ZSS uses the previous version of ZIS?

No token is returned.

Reply:
my understanding is the zss portion should be moved into a plugin to preserve flexibility
it's not hard to do considering the difference between zss plugins and zss services are boilerplate that packages a service. if you have a plugin, we can get it built using our existing plugin pipelines and i can record it quickly here https://github.com/zowe/zowe-install-packaging/tree/master/files/zlux/config/plugins

Reply:
https://github.com/zowe/explorer-ip is a zss plugin currently shipped in zowe server releases.

Reply:
I suppose this is just fixing existing functionality in ZIS, ZSS, and APIML. When one of the pieces is missing, nothing will work. So does it make sense to deliver it as a plugin?

Reply:
That original service wasn't designed correctly I suppose. It doesn't have any versioning. That's why extending it is problematic.

I can see the following options:
* Make another core service
* Change that existing core service to have another function code
** The new service would only append new field to `AuthServiceParmList`
** The new service would add a version field which would be checked when the new function code is used
* Make a ZIS plugin

My preference would be a plugin:
* It's easier to isolate any security issue
* Can be updated separately

Of course, it will be more problematic to create a plugin because it will entail adding an additional artifact, updating the doc, etc. Also the documentation is lacking.

Perhaps we need to have a discussion.

Reply:
I don't mind any good solution but I would need to have some plan. It is blocking the merging of https://github.com/zowe/api-layer/pull/2083 and this is blocking some additional work.

Reply:
* We shouldn't be using the existing function code (`ZIS_AUTH_SERVICE_PARMLIST_FC_VERIFY_PASSWORD`) for getting IDT in the auth service, because it seems like a different function than verifying a password?
* Versioning must be present in the parameter list to avoid issues when different versions of the APIs are used
* Ideally, this would be a plugin (not sure what this entails with regards to timing, probably not in favor of this feature: that'd be the first public plugin and it'd need to be integrated into packaging, installer, config etc.)

Items 1 and 2 are a must. @1000TurquoisePogs , @JoeNemo what do you think about item 3? I'd personally move all functionality except for the print, dump and router services to plugins, but that probably requires a major version change.